src/core/components-ipfsx/add/index.js

```diff
@@ -0,0 +1,115 @@
+'use strict'
+
+const importer = require('ipfs-unixfs-importer')
+const normaliseAddInput = require('ipfs-utils/src/files/normalise-input')
+const { parseChunkerString } = require('./utils')
+const pipe = require('it-pipe')
+const { withFirstAndLast } = require('../../utils')
+
+module.exports = ({ ipld, dag, gcLock, preload, pin, constructorOptions }) => {
+  return withFirstAndLast(async function * add (source, options) {
+    options = options || {}
+
+    const opts = {
+      shardSplitThreshold: constructorOptions.EXPERIMENTAL.sharding ? 1000 : Infinity,
+      ...options,
+      ...parseChunkerString(options.chunker)
+    }
+
+    // CID v0 is for multihashes encoded with sha2-256
+    if (opts.hashAlg && opts.cidVersion !== 1) {
+      opts.cidVersion = 1
+    }
+
+    if (opts.progress) {
+      let total = 0
+      const prog = opts.progress
+
+      opts.progress = (bytes) => {
+        total += bytes
+        prog(total)
+      }
+    }
+
+    const iterator = pipe(
+      normaliseAddInput(source),
+      source => importer(source, ipld, opts),
+      transformFile(dag, opts),
+      preloadFile(preload, opts),
+      pinFile(pin, opts)
+    )
```

Comment:
I _really_ like pipelines like this one. 
Easy to understand what needs to happen just by scanning only a few lines without going too deep into details. 
Doing more of this would make it easier to onboard new contributors to the codebase.

Reply:
me too!